JavaScript/testing.md

```diff
@@ -0,0 +1,109 @@
+
+<a name="testing-for-data-flow"></a>
+
+### Testing for Data Flow
+
+The best way to ensure great and defensive data flows in the app are not tests, they're **types**. A great type system will prevent more data errors than any test suite. This is for a few reasons.
+
+**Types** establish explicit contractual obligations between functions. If something is a `string` or `undefined`, it is known to both the developer and to the type system. If we try to do something dangerous when that `string` is possibly `undefined`, the type system will alert us, sometimes even preventing the code from compiling. This forces us to code defensively at every step of our application, preventing us from trying to manipulate data in a way that will cause errors.
+
+It also forces us to think more critically about the data and functions we create. Eventually, making sure our code is defensive will become second nature since we will learn how to avoid all sorts of type errors. It will prevent spelling errors and errors like trying to access a variable of a parent object that is possibly `undefined`. It will also aid heavily in ensuring that our reducers are functioning as expected, since we can type reducers to always return a certain type of state - `HomeState` for example. Anything that is returned within that reducer will need to fit that existing type.
+
+If any code is refactored, it will immediately point out errors with accessing incorrect or stale data. For example, if we were to refactor a slice of our state, any `mapStateToProps` function that pull from it will be aware of the new state and error if we are trying to pull state that does not exist. Similarly, any `prop` we pass will now have an updated signature, preventing any runtime errors, like converting an array into an object and trying to map over that object. Refactors like this are frequent once a project starts reaching a few dozen components.
+
+<a name="testing-for-accidental-code-changes"></a>
+
+### Testing for Accidental Code Changes
+
+The two main defenses against accidental code changes are **Code Reviews** and [**Snapshot Tests**](https://jestjs.io/docs/en/snapshot-testing). Code Reviewers can spot accidental changes, which is they are so important. **Snapshot Tests** can also point out to the developer that their code has changed in a file or module that they were not supposed to be editing. For example, if we change some of the login logic and a snapshot test fails in our search component, we've probably done something wrong or did a code change where we shouldn't have.
```

Comment:
Linting as a part of the test suite could also catch calls to `console` methods and the `debugger`.

Reply:
ah true. my only problem with linting is that sometimes `console.logs` are put in as placeholders for errors that we don't really want to surface to users, but may want to surface to developers. I could see us centralizing that as a service though and ignoring the single line that the service will call. Even if that's the case, should we error when a `console.log` is written?

Just trying to think if there's a way to change it between develop and when we're actually running it in CI. maybe just passing in a param that strictly enforces it during CI? I'm guessing we can then handle debuggers in the same way

Reply:
Contextually-speaking, console methods should only remain in the code if they're only called in local development or even staging environments, unless the console is how the primary user is interacting with the product (like in Node.js).

Alternatively speaking, if there's a way to build FE JS in such a way that `console`'s of different severities can be included or excluded from the compiled JS (and we can define that on build-time) that would be perfect (no linting required).